src/LiveComponent/assets/src/url_utils.ts

```diff
@@ -0,0 +1,169 @@
+    return true;
+}
+
+/**
+ * Converts JavaScript data to bracketed query string notation.
+ *
+ * Input: `{ items: [['foo']] }`
+ *
+ * Output: `"items[0][0]=foo"`
+ */
+function toQueryString(data: any) {
+    const buildQueryStringEntries = (data: { [p: string]: any }, entries: any = {}, baseKey = '') => {
+        Object.entries(data).forEach(([iKey, iValue]) => {
+            const key = baseKey === '' ? iKey : `${baseKey}[${iKey}]`;
+
+            if ('' === baseKey && isValueEmpty(iValue)) {
+                // Top level empty parameter
+                entries[key] = '';
+            } else if (null !== iValue) {
+                if (typeof iValue === 'object') {
+                    // Non-empty object/array process
+                    entries = { ...entries, ...buildQueryStringEntries(iValue, entries, key) };
+                } else {
+                    // Scalar value
+                    entries[key] = encodeURIComponent(iValue)
+                        .replace(/%20/g, '+') // Conform to RFC1738
+                        .replace(/%2C/g, ',');
+                }
+            }
+        });
+
+        return entries;
+    };
+
+    const entries = buildQueryStringEntries(data);
+
+    return Object.entries(entries)
+        .map(([key, value]) => `${key}=${value}`)
+        .join('&');
+}
+
+/**
+ * Converts bracketed query string notation to JavaScript data.
+ *
+ * Input: `"items[0][0]=foo"`
+ *
+ * Output: `{ items: [['foo']] }`
+ */
+function fromQueryString(search: string) {
+    search = search.replace('?', '');
+
+    if (search === '') return {};
+
+    const insertDotNotatedValueIntoData = (key: string, value: any, data: any) => {
+        const [first, second, ...rest] = key.split('.');
+
+        // We're at a leaf node, let's make the assigment...
+        if (!second) return (data[key] = value);
+
+        // This is where we fill in empty arrays/objects along the way to the assigment...
+        if (data[first] === undefined) {
+            data[first] = Number.isNaN(Number.parseInt(second)) ? {} : [];
+        }
+
+        // Keep deferring assignment until the full key is built up...
+        insertDotNotatedValueIntoData([second, ...rest].join('.'), value, data[first]);
+    };
+
+    const entries = search.split('&').map((i) => i.split('='));
+
+    const data: any = {};
+
+    entries.forEach(([key, value]) => {
+        value = decodeURIComponent(value.replace(/\+/g, '%20'));
+
+        if (!key.includes('[')) {
+            data[key] = value;
+        } else {
+            // Skip empty nested data
+            if ('' === value) return;
+
+            // Convert to dot notation because it's easier...
+            const dotNotatedKey = key.replace(/\[/g, '.').replace(/]/g, '');
+
+            insertDotNotatedValueIntoData(dotNotatedKey, value, data);
+        }
+    });
+
+    return data;
+}
+
+/**
+ * Wraps a URL to manage search parameters with common map functions.
+ */
+export class UrlUtils extends URL {
```

Comment:
If i understand correctly, the URL is only changed after a replace (from the backend)..

Do we really need to duplicate all that URL / query logic in the front ? 

Is it not possible to send the URL from the back-end ? I must miss something here I think :)

Reply:
Well, it _might_ be possible, but I'd prefer to keep the implementation on frontend side.

I don't think the URL / query logic is really duplicated. I'd rather say it's split by responsibility. The frontend is responsible for the current URL manipulation. I don't know why we would like to delegate the computing to the server, since there is all the required material to do it on client side. It also avoids to pass the current URL as an attribute in each Live Component request.

Reply:
Just a tentative to avoid adding weight to the JS code but i understand your point.